### DISCUSSION TITLE
ansible 질문 있습니다.

### DISCUSSION THREAD

**Discussion by jangzzi** [tedilabs/fastcampus-devops#36] [Q&A] [answered]
ansible 시작할 때 ec2 머신 4개 생성하는 부분에서 sg 생성하는데요.
아래 이유로 생성이 안됩니다. 어느 부분을 수정해야하나요?
리소스를 보면 회사 테라폼 백엔드 경로 같은데 접근이 안되서 에러나는거 같네요.
코드 어떻게 수정하면 바로 진행할 수 있을까요?

astcampus@DESKTOP-GRB1BMM:~/ansible/ec2-instance$ pwd
/home/fastcampus/ansible/ec2-instance

fastcampus@DESKTOP-GRB1BMM:~/ansible/ec2-instance$ ls
config.yaml  files       remote-states.tf    terraform.tf  versions.tf
ec2.tf       outputs.tf  security-groups.tf  variables.tf

fastcampus@DESKTOP-GRB1BMM:~/ansible/ec2-instance$ tf init
Initializing modules...
╷
│ Error: Module not found
│
│ Module "sg__ssh" (from security-groups.tf:1) cannot be found in the module registry at
│ app.terraform.io.
╵
╷
│ Error: Module not found
│
│ Module "sg__ssh" (from security-groups.tf:1) cannot be found in the module registry at
│ app.terraform.io.

fastcampus@DESKTOP-GRB1BMM:~/ansible/ec2-instance$ tf apply
│ Error: Module not installed
│
│   on security-groups.tf line 1:
│    1: module "sg__ssh" {
│
│ This module is not yet installed. Run "terraform init" to install all modules required by this
│ configuration.

**Comment by posquit0** [comment, marked as answer]
안녕하세요! 이슈 제보해주셔서 정말 감사드립니다. 😢 
제가 미처 놓친 부분이 있었네요... ㅎㅎ;
최신 코드로 다시 받아주시면 공개 레포지토리 상의 `security-group` 모듈을 받아오게 됩니다.

다음과 같이 `module`의 `source`를 수정하였습니다. :)

```tf
source  = "tedilabs/network/aws//modules/security-group"
```

**Reply by jangzzi** [reply]
위 부분은 해결됬는데, 이제 remote-states에서 또 에러가 발생하네요.

│ Error: Unsupported attribute
│
│   on remote-states.tf line 5, in locals:
│    5:     vpc           = local.remote_states["network"].vpc
│     ├────────────────
│     │ local.remote_states["network"] is object with 3 attributes
│
│ This object does not have an attribute named "vpc".
╵
╷
│ Error: Unsupported attribute
│
│   on remote-states.tf line 6, in locals:
│    6:     subnet_groups = local.remote_states["network"].subnet_groups
│     ├────────────────
│     │ local.remote_states["network"] is object with 3 attributes
│
│ This object does not have an attribute named "subnet_groups".

**Reply by seank01** [reply]
저도 윗분과 같은 상황이네요, 확인 좀 부탁드립니다

**Reply by posquit0** [reply]
안녕하세요~ 답변이 늦었습니다.

https://github.com/tedilabs/fastcampus-devops/blob/main/2-ansible/ec2-instance/config.yaml 파일을 확인해보면 `network` 정보를 `remote_states`에 설장한 내용을 확인할 수 있습니다.

```
remote_states:
  "network":
    organization: "fastcampus-devops"
    workspace: "terraform-lab-network"
```

해당 코드는 `remote-states.tf`에서 실제로 아래와 같이 참조되고 있습니다.

```tf
data "terraform_remote_state" "this" {
  for_each = local.config.remote_states

  backend = "remote"

  config = {
    organization = each.value.organization
    workspaces = {
      name = each.value.workspace
    }
  }
}
```

이 내용은 테라폼 파트에서 테라폼 클라우드를 이용하여 `remote-states` 기능을 활용하는 부분에 대해서 공유를 드렸던 내용인데요!
`network`를 구성하는 테라폼 코드를 테라폼 클라우드 상태 저장소를 이용하여 `terraform apply`한 경우, 해당 네트워크 정보를 다른 워크스페이스, 여기서는 `ec2-instance` 워크스페이스에서 참조하기 위해서 `terraform_remote_state` 데이터 소스에서 `backend = "remote"`를 이용하여 데이터를 조회하여야 합니다.

```
remote_states:
  "network":
    organization: "fastcampus-devops"
    workspace: "terraform-lab-network"
```

여기서 `organization: "fastcampus-devops"`는 제가 실습 목적으로 사용한 조직 명이라, 수강자 분께서 사용하신 `organization` 명으로 수정하시고 `terraform login`이 된 상태(테라폼 클라우드와 인증된 통신을 하기 위함)에서 `terraform apply`를 수행하신다면 이슈가 발생하지 않을 것으로 보입니다.


### 관련 자료

- https://www.terraform.io/cloud-docs/workspaces/state#data-source-configuration
- https://www.terraform.io/cli/commands/login

**Reply by seank01** [reply]
메일까지 답변 달아주시고, 결론적으로 해결 잘 되었습니다, 기존까지 다른 모든 실습 내용의 문제가 없었기 때문에 Backend (Remote) 등으 설정 문제는 아니었구여, 여러번 응용 변경하여 실행 하던중 발생된 문제인듯 하구여, 기존 테라폼 설정 파일들 전부 삭제하고 실행하니 문제없이 잘 되었습니다. 감사드립니다
